File: recycler-decoration/src/main/java/com/android/oleksandrpriadko/recycler_decoration/ItemDecorationMargin.java
package com.android.oleksandrpriadko.recycler_decoration;

import android.graphics.Rect;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

public class ItemDecorationMargin extends RecyclerView.ItemDecoration {

    private int mLeftMargin;
    private int mTopMargin;
    private int mRightMargin;
    private int mBottomMargin;
    private boolean mFirstItemMargin;
    private boolean mLastItemMargin;
    private int mLinearLayoutManagerOrientation;

    public ItemDecorationMargin(int leftMargin,
                                int topMargin,
                                int rightMargin,
                                int bottomMargin,
                                boolean firstItemMargin,
                                boolean lastItemMargin,
                                int linearLayoutManagerOrientation) {

        this.mLeftMargin = leftMargin;
        this.mTopMargin = topMargin;
        this.mRightMargin = rightMargin;
        this.mBottomMargin = bottomMargin;
        this.mFirstItemMargin = firstItemMargin;
        this.mLastItemMargin = lastItemMargin;
        this.mLinearLayoutManagerOrientation = linearLayoutManagerOrientation;
    }

    public ItemDecorationMargin(int margin,
                                boolean firstItemMargin,
                                boolean lastItemMargin,
                                int linearLayoutManagerOrientation) {

        this(margin,
            margin,
            margin,
            margin,
            firstItemMargin,
            lastItemMargin,
            linearLayoutManagerOrientation);
    }

    @Override
    public void getItemOffsets(Rect outRect,
                               View view,
                               RecyclerView parent,
                               RecyclerView.State state) {
        int position = parent.getChildAdapterPosition(view);
        boolean isFirstItem = position == 0;
        boolean isLastItem = position == position - 1;

        if (mLinearLayoutManagerOrientation == LinearLayoutManager.HORIZONTAL) {
            defineOffsetHorizontalOrientation(outRect, isFirstItem, isLastItem);
        } else if (mLinearLayoutManagerOrientation == LinearLayoutManager.VERTICAL) {
            defineOffsetVerticalOrientation(outRect, isFirstItem, isLastItem);
        }
    }

    private void defineOffsetVerticalOrientation(Rect outRect,
                                                 boolean isFirstItem,
                                                 boolean isLastItem) {
        outRect.left = mLeftMargin;
        outRect.right = mRightMargin;

        if (isFirstItem) {
            if (mFirstItemMargin) {
                outRect.top = mTopMargin;
            }
            outRect.bottom = mBottomMargin;
        } else if (isLastItem) {
            if (mLastItemMargin) {
                outRect.bottom = mBottomMargin;
            }
        } else {
            outRect.bottom = mBottomMargin;
        }
    }

    private void defineOffsetHorizontalOrientation(Rect outRect,
                                                   boolean isFirstItem,
                                                   boolean isLastItem) {
        outRect.top = mTopMargin;
        outRect.bottom = mBottomMargin;

        if (isFirstItem) {
            if (mFirstItemMargin) {
                outRect.left = mLeftMargin;
            }
            outRect.right = mRightMargin;
        } else if (isLastItem) {
            if (mLastItemMargin) {
                outRect.right = mRightMargin;
            }
        }
    }
}

File: recycler-decoration/src/main/java/com/android/oleksandrpriadko/recycler_decoration/ItemDecorationMargin.kt
package com.android.oleksandrpriadko.recycler_decoration

import android.graphics.Rect
import android.view.View

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class ItemDecorationMargin(private val leftMargin: Int,
                           private val topMargin: Int,
                           private val rightMargin: Int,
                           private val bottomMargin: Int,
                           private val firstItemMargin: Boolean,
                           private val lastItemMargin: Boolean,
                           private val linearLayoutManagerOrientation: Int)
    : RecyclerView.ItemDecoration() {

    constructor(margin: Int,
                firstItemMargin: Boolean,
                lastItemMargin: Boolean,
                linearLayoutManagerOrientation: Int)
            : this(
            margin,
            margin,
            margin,
            margin,
            firstItemMargin,
            lastItemMargin,
            linearLayoutManagerOrientation)

    override fun getItemOffsets(outRect: Rect,
                                view: View,
                                parent: RecyclerView,
                                state: RecyclerView.State) {
        val position = parent.getChildAdapterPosition(view)
        val isFirstItem = position == 0
        val isLastItem = position == parent.adapter!!.itemCount - 1

        if (linearLayoutManagerOrientation == LinearLayoutManager.HORIZONTAL) {
            defineOffsetHorizontalOrientation(outRect, isFirstItem, isLastItem)
        } else if (linearLayoutManagerOrientation == LinearLayoutManager.VERTICAL) {
            defineOffsetVerticalOrientation(outRect, isFirstItem, isLastItem)
        }
    }

    private fun defineOffsetVerticalOrientation(outRect: Rect,
                                                isFirstItem: Boolean,
                                                isLastItem: Boolean) {
        outRect.left = leftMargin
        outRect.right = rightMargin

        if (isFirstItem) {
            if (firstItemMargin) {
                outRect.top = topMargin
            }
            outRect.bottom = bottomMargin
        } else if (isLastItem) {
            if (lastItemMargin) {
                outRect.bottom = bottomMargin
            }
        } else {
            outRect.bottom = bottomMargin
        }
    }

    private fun defineOffsetHorizontalOrientation(outRect: Rect,
                                                  isFirstItem: Boolean,
                                                  isLastItem: Boolean) {
        outRect.top = topMargin
        outRect.bottom = bottomMargin

        if (isFirstItem) {
            if (firstItemMargin) {
                outRect.left = leftMargin
            }
            outRect.right = rightMargin
        } else if (isLastItem) {
            if (lastItemMargin) {
                outRect.right = rightMargin
            }
        } else {
            outRect.right = rightMargin
        }
    }
}
